### DISCUSSION TITLE
Building Plan for new in-game Items: Combining X2 and 38A Elements?

### DISCUSSION THREAD

**Discussion by 0lhi** [Wohlstand/TheXTech#239] [Planning]
EDIT: The old Combination Idea is abandoned in favor of:

- Hardcoding the 38A-Elements into TheXTech.
- Allow Lua-Scripts to overwrite new hardcoded content.

See below for more.

**Comment by Alucard648** [comment]
I think, we should start with implementing proper scripting support and then follow the same roadmap SMBX2 did, while also stramlining some key technical aspects, like goddamn powerup array. And Editor is well out of date and also needs rewriting, complete with Java Script plugin support. Some Lua scripts are almost unuseable without WYSiWYG editor or such, like SMWMap.lua.

**Reply by Wohlstand** [reply]
> powerup array

The case, here, at TheXTech, it's easy to extend it by just global variable editing in the code, I did the same with sections array (that why 201 sections is supported here)

**Reply by Wohlstand** [reply]
Anyway, forgot to tell, this is just planning for a future, right now there are core parts in work

**Reply by LoveBodhi** [reply]
So yes...

**Comment by LoveBodhi** [comment]
My FTW have more extended elements so will be used in the future...

**Comment by vitzie629** [comment]
What would happen if every single element from SMB1 - SMW was added (the elements from SMB1 - SMW that don't exist in SMBX) and would be very similar to the fake SMBX v1.4 (fake screenshots using MS-Paint by DuckJohnn according to this page: https://wohlsoft.ru/pgewiki/Branches_of_Super_Mario_Bros._X#Other_projects)?

**Reply by LoveBodhi** [reply]
That meme is like my future season-based textures.

**Reply by vitzie629** [reply]
> That meme is like my future season-based textures.

What are you talking about?

**Reply by LoveBodhi** [reply]
My "Fairy Tale World" asset!

**Reply by vitzie629** [reply]
> My "Fairy Tale World" asset!

That’s right!

**Comment by ChristianSilvermoon** [comment]
I'd personally hope for all features of both 2 and 38A, but I know both implement some of the same things and sometimes they work slightly differently. I also know that 38A's physics are different.

But I know that that's a tall order filled with about a million technical challenges that I am probably not aware of to doing such a thing... ID conflicts I imagine are likely a problem.

Perhaps in the case of duplicate objects, some kind of flag could be used to determine if that object should behave SMBX2 or SMBX38A style and the game could automatically set such flags (if they're not in the level file) at runtime to retain file compatibility?

**Reply by 0lhi** [reply]
> I'd personally hope for all features of both 2 and 38A, but I know both implement some of the same things and sometimes they work slightly differently.

Since 38A is a different engine with no available sources, I think we should prioritize accuracy for X2, and settle with getting 38A stuff to a playable state.

> But I know that that's a tall order filled with about a million technical challenges that I am probably not aware of to doing such a thing... ID conflicts I imagine are likely a problem.

Yeah, that's what I hope to resolve with the building plan. Add the stuff from X2 and 38A to an ID list that resembles theirs as closely as possible, but also works as it's own thing that doesn't restrict itself to the flaws of the originals.

With that approach, the translation API that Wohlstand mentioned would be easier to make, too.

**Reply by LoveBodhi** [reply]
Unlike these, my FTW uses custom sorting system instead (Init position still retain, 751 - 1000).

**Comment by 0lhi** [comment]
After taking a look at the additions of both Engines, I was surprised to see that 38A adds way less than X2: [38A+X2.zip](https://github.com/Wohlstand/TheXTech/files/9277197/38A%2BX2.zip)
 
 Only 65 new NPCs and 10 new BGOs. No new Blocks at all! Unlike SMBX2, the 38A additions are all hardcoded. After speaking with @Wohlstand and @ds-sloth, our idea how to go on now is as follows:

**1. Hardcode the 75 38A-Additions into TheXTech.**
**2. Allow Lua-Scripts (discussed in #129) to overwrite hardcoded NPCs (≥293), BGOs (≥191) & Blocks (≥639).**
**3. Have Lua-Scripts be handled exclusively by Asset Packs (discussed in #291).**

**Things to consider:**

- npc-293-300 will need a Special Value to preserve the Dummies for Vanilla episodes that use them. Imo, the 38A NPCs should be the default, while the Dummies should be a fallback for Level-Files targeting SMBX64 or lower. See #185 and #163. Additionally, a Compat-setting may be made.
- Some of the NPCs are PowerUps, which also entails changes for the Characters.
- NPC-Functions will likely need to be rewritten. @ds-sloth and @Wohlstand can add more details.

**What this implies for 38A-Support:**

- TheXTech and 38A will have a shared base of NPCs, BGOs and Blocks. Many 38A-Episodes will get into a playable state, though 100% accuracy is not a priority.
- Certain features exclusive to 38A may be added if it's convenient to do.
- TeaScript and support for encrypted 38A-Episodes will not be added. Episodes making use of those need to be hacked and ported.

**What this implies for X2-Support:**

- The NPCs, BGOs and Blocks of SMBX2 will generally not be hardcoded into TheXTech. Exceptions may be made for individual situations if there are good reasons (increasing performance or high demand), but that would only apply to very few select cases.
- SMBX2-Preservation (in form of an X2-Asset Pack) and the maintenance of new Asset Packs would be outsourced to the Community. The priority of TheXTech-development would be to create a good scripting environment that allows expansion, without aiming to expand themselves.

If you have Feedback, Suggestions or Objections, we are eager to hear!

**Reply by ds-sloth** [reply]
While this was @0lhi's idea, from my perspective it seems like a very reasonable strategy.

By keeping the Lua-incompatible NPCs hardcoded, we're able to limit scope bloat on the scripting API and keep reasonable limitations for what can be scripted. (For instance, we can prevent NPCs from accessing each other's internal state, and push scripts to use high-performance hardcoded API functions that make use of the game's quadtrees instead of low-performance loops in Lua itself.) Additionally, by keeping the SMBX Type IDs hardcoded, we will be able to add scripted custom NPCs with practically no performance impact on SMBX64 levels.

Of course, the implementation will be difficult, and depending on how the timeline proceeds, it is very possible that scripting support will be implemented before all of the SMBX 38A NPCs are done.

**Reply by Wohlstand** [reply]
I gonna try the basic scripting thing after the 1.3.6 release and multi-res merge, so, with a minimalistic base, we could make the API that should handle all our needs. And, re-implementing 38A AIs will take some time, just need to build a list, and try to implement each by the "black box" method. But again, before to start implementing any, I should build the skeleton of API overriding function pointers for the future.

**Reply by ds-sloth** [reply]
That sounds perfect. When you have the function pointer and attribute skeleton, I'd like to review it for performance and any possible additional features.

**Reply by LoveBodhi** [reply]
However, my Nostalgic Paradise uses custom standard, instead of hardcoded or X2/38A standard.

**Reply by ds-sloth** [reply]
If you prefer to not use any of the SMBX64 NPCs, you don't need to use any of them. You could just skip those NPC IDs in your NPC lists.

**Reply by LoveBodhi** [reply]
Only a small part of NPC's aren't SMBX64 for my assets.

**Reply by ds-sloth** [reply]
Sorry, maybe I didn't understand your note perfectly. It sounds like you are saying that you will use mostly SMBX64 NPCs, and some custom NPCs of your own design. That should be fine. You could use the NPC Type slots >= 293 for your custom NPCs, and then include whichever SMBX64 NPCs you want to in your standard as well.

**Reply by ChristianSilvermoon** [reply]
A bit different from my initial thoughts

I originally thought that the idea was to hardcode *both* and where possible merge identical elements (because their are a few) and have a flag of some kind to switch a particular object from it's X2 behavior to it's 38A Behavior... as long as performance is reasonable, I guess the "Hardcode 38A/Softcode X2" approach sounds okay.

But some questions I now have are:
* What does this mean for unified content? Will putting 38A and X2 elements in the same level be possible?
* 38A has a lighting system only activate-able via TeaScript, however, I see no particular reason that should be necessary. Could we enable a similar/the same lighting system in TheXTech, and possibly share the logic with X2's lighting thingy?
* Wall Jumping is activated *per level* with an Event in 38A... I would propose this be made activatable in the same way for compatibility's sake, but... this is an obnoxious way to make an entire episode with it, could we possibly make some kind of flag settable for the whole episode instead?

**Reply by LoveBodhi** [reply]
So will be a mode switch?

**Reply by LoveBodhi** [reply]
But also, 38A renders via DirectX and X2 renders via OpenGL.

**Reply by ds-sloth** [reply]
> Will putting 38A and X2 elements in the same level be possible?

I think someone would need to make a different asset pack that includes the X2 NPCs with higher Type IDs, above the 38A ones. Is that problematic in some way?

> 38A has a lighting system...

I don't think we're actively planning a lighting system right now, but we'd probably make our own, which would attempt to mimic the SMBX2 lighting system more closely than the 38A one.

> 38A renders via DirectX...

We are not going to use DirectX, and I think it's very unlikely that we could directly copy the lighting systems of either engine (but more possible with SMBX2).

> Wall Jumping is activated per level with an Event in 38A

If we implement Wall Jumping, we would certainly also make a config/compat option for it.

**Reply by 0lhi** [reply]
> I think someone would need to make a different asset pack that includes the X2 NPCs with higher Type IDs, above the 38A ones. Is that problematic in some way?

@ds-sloth One thing I could imagine being helpful is if Lua-Scripts get the ability to say assign themselves to a hardcoded NPC. So if a Level for the X2 Asset Pack wants to make use of a 38A-PowerUp (npc-323 Penguin), but that is already overwritten by an X2 Enemy (Stretch Boo), npc-600.lua could say ```I'm actually npc-323```, giving access to the PowerUp. Would that be possible?

> Could we enable a similar/the same lighting system in TheXTech, and possibly share the logic with X2's lighting thingy?

Probably best to discuss at #131.

> Wall Jumping is activated per level with an Event in 38A... I would propose this be made activatable in the same way for compatibility's sake, but... this is an obnoxious way to make an entire episode with it, could we possibly make some kind of flag settable for the whole episode instead?

Oh, does 38A have hardcoded Wall-Jumps? @ChristianSilvermoon

**Reply by ChristianSilvermoon** [reply]
> Oh, does 38A have hardcoded Wall-Jumps? @ChristianSilvermoon
It does, actually, but you have to Enable the functionality with an on-start Event for some reason ^^

I'm going to make a big list of everything I can find that's 38A Exclusive or different to help out I've already got one going in another reply x3

**Reply by 0lhi** [reply]
That will certainly be helpful :3

**Reply by ds-sloth** [reply]
> Would it be possible to assign to hardcoded NPC?

It would be downright easy if we manage to reimplement all of the NPCs in the function-table style @Wohlstand described (and you would probably be able to partially override only one of their functions), but if not, it will be at least simple to assign to the 38A NPCs.

**Comment by ChristianSilvermoon** [comment]
**List moved to #379**

----

**Last Updated:** Mon 08 Aug 2022 05:31:02 AM UTC [...]

## Thoughts for Substituting Certain 38A Features Without TeaScript
Various parts of 38A can react to conditional statements using variables.

It should be possible to avoid any need for TeaScript on TheXTech's side by allowing user-defined variables to be read and set via Events.

In 38A's TeaScript, it is also possible to use certain internal variables such as The Star Count... I would propose making it possible to trigger events based on Star Counts, Player Power Up State, and Mount as a substitute.

Certain TeaScript function such as `LSpin` could also easily be replicated by extending events.

**Reply by 0lhi** [reply]
Amazing Work! Very insightful ^-^

> A block-ID.txt flag (assuming those are a thing) to set whether something can be wall jumped off of or not?

Is block-ID.txt a thing in 38A? They are not in TheXTech. Moondust has block-ID.ini I think.

> Customizable Player Physics, and a compat flag to default to 38A physics, or the best approximation of them we can produce

Imo, compat flag should only determine abilities (wall jump, double jump, midair spin etc.). Physics should be exclusively determined by a Players control profile. SMBX should be the canonical main, while 38A and SMW (see #157) would be secondary options.

> Set Viewport (What does THIS do?)

Makes me curious as well :o I probably should setup a VM to do 38A Experiments.

**Reply by LoveBodhi** [reply]
> Is block-ID.txt a thing in 38A? They are not in TheXTech. Moondust has block-ID.ini I think.

SMBX2

**Reply by ChristianSilvermoon** [reply]
> Imo, compat flag should only determine abilities (wall jump, double jump, midair spin etc.). Physics should be exclusively determined by a Players control profile. SMBX should be the canonical main, while 38A and SMW (see #157) would be secondary options.

I specifically mention this for compatibility's sake. 38A's physics differ from 1.3's and playing the levels with default 1.3 physics can be game breaking, and thus, would be a problem of compatibility for those levels/episodes.

Handling physics only at the Control Profile would be not very good for Episodes, because often times level design will be heavily based around a specific set of physics. Adding the option to Control Profiles I can agree with, but Episodes should _absolutely_ be able to designate and/or override them to open more opportunities for unique and custom experiences as well as make sure the levels are actually _playable_.

EDIT: Additionally, Wall jumps are a core 38A feature, enabled or disabled by events. I'm merely suggesting a compat flag that sets the default state to "on" instead of the default state being "off" so that you don't have to set up the same event in 50 different levels for a new episode that allows Wall Jumping.

EDIT2: If we decide to go with this, should I create a sort of Master Issue to track 38A Elements added  to TheXTech?

**Reply by 0lhi** [reply]
> 38A's physics differ from 1.3's and playing the levels with default 1.3 physics can be game breaking

Can it? Abilities aside (which should be configurable via compat), I don't see a scenario where you can't do something with the more precise SMBX mechanics that you can do with 38A's mechanics. If you have an example, feel free to share. Otherwise, I think 38A-Physics are only worth considering as an completely optional extra at best. I'd be perfectly fine with not having them at all.

> Adding the option to Control Profiles I can agree with, but Episodes should absolutely be able to designate and/or override them to open more opportunities for unique and custom experiences as well as make sure the levels are actually playable.

If episodes override them users will just be able to override again with the compat rewrite. Players will always get the last word.

The reason it shouldn't be done via compat is that it doesn't distinguish between Player 1 and 2. So, it should be left with the Profile.

What Episode Makers can and should do is suggest a set of mechanics if it's needed. As long as Physics can be changed on the fly, playability won't be harmed and the opportunities will be there. And they'll actually be opportunities rather than impositions.

> EDIT2: If we decide to go with this, should I create a sort of Master Issue to track 38A Elements added to TheXTech?

I actually wanted to do this like next week or so, but you are clearly more qualified :P

I'd still say let's wait until Sunday, give some time for others to give feedback.

**Reply by ds-sloth** [reply]
0lhi, I see that you really want the two players to be able to select their preferred physics individually. But the controls profiles settings have to do with *controls*, not gameplay. Also, the controls profiles settings are designed to be predictable and totally enforce the player’s intent. If it’s set up with control profiles, that would completely prevent episode creators from even suggesting a physics other than a player’s preferred choice.

Instead, we should use the compat rewrite standard where the episode creator can override the player’s global settings, but the player can in turn override that. Would per-character physics selection in the compatibility settings be a good compromise here?

(This whole point may be moot because I also think we might never implement 38A physics.)

**Reply by LoveBodhi** [reply]
I see, it should be per-character physic...

**Reply by 0lhi** [reply]
> Would per-character physics selection in the compatibility settings be a good compromise here?

Not quite since sometimes the same character is used by multiple players, but per-player selection could work.

> (This whole point may be moot because I also think we might never implement 38A physics.)

We also need to keep in mind that LunaLua Scrips can introduce Physics Changes, too. I don't know about TeaScript.

**Reply by ChristianSilvermoon** [reply]
> This whole point may be moot because I also think we might never implement 38A physics.

What I'm actually suggesting is an easy way for Episode Designers to customize character physics in detail, but to implement optional easy "Physics Profiles" so that should an Episode Designer want custom physics they can just slap in "38A" or "SMW" rather than typing out every physics change themselves, which would be very complex and less accessible.

> Players will always get the last word.

I understand where you're coming from, but I've got to disagree that it's for the best. Level design is **almost always** tightly coupled with physics, and in fact this is even a part of Nintendo's core design philosophy. Levels are designed around a character's movement and capabilities.

If one set of physics lowers Jump Height, the player may be unable to clear a gap they otherwise would have been able to clear.

If another set lowers top run speed, it may make some sections impossible to survive.

Something as essential as how the character controls is probably best left to the designer... I'm not opposed to giving the player options (I LOVE options), but this is an option that may **frequently** lead to unfair feeling deaths and confusion.

If allowing the player to override physics is available, it should come with a prominent caution advisory.

**Reply by 0lhi** [reply]
> EDIT2: If we decide to go with this, should I create a sort of Master Issue to track 38A Elements added to TheXTech?

@ChristianSilvermoon Since no one objected, I think you can make the Master Issue now if you're still up for it. Feel free to list any 38A feature that you can find. Not everything will be added, but it will be useful to have a list that fully documents what is added, and what was decided against.

**Comment by LoveBodhi** [comment]
38A have click event as well.

**Comment by LoveBodhi** [comment]
It also combines issues #94 , #114 and #324 .

**Comment by LoveBodhi** [comment]
We might need to create a "Mode Switch" for non-SMBX64 objects:

- Mode 0: Custom (100% User-defined)
- Mode 1: SMBX-38A Standard
- Mode 2: SMBX2 Standard

**Comment by LoveBodhi** [comment]
I think, if there's an "NPC / Block AI Editor" system, that will be fine...

Example:

- After n Frames
- Do AI 1
- After n Frames
- Do AI 2
- etc...

Or the AI can be looped back.

**Comment by LoveBodhi** [comment]
I forgot there's a forum poll https://wohlsoft.ru/forum/viewtopic.php?f=94&t=4456
